messaging/src/main/java/org/axonframework/eventhandling/deadletter/jpa/JpaSequencedDeadLetterQueue.java

```diff
@@ -0,0 +1,772 @@
+import org.axonframework.serialization.Serializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandles;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+import javax.annotation.Nonnull;
+import javax.persistence.EntityManager;
+import javax.persistence.NoResultException;
+
+import static org.axonframework.common.BuilderUtils.*;
+
+/**
+ * JPA-backed implementation of the {@link SequencedDeadLetterQueue}, used for storing dead letters containing
+ * {@link EventMessage Eventmessages} durably as a {@link DeadLetterEntry}.
+ * <p>
+ * Keeps the insertion order intact by saving an incremented index within each unique sequence, backed by the
+ * {@link DeadLetterEntry#getIndex()} property. Each sequence is uniquely identified by the sequence identifier, stored
+ * in the {@link DeadLetterEntry#getSequenceIdentifier()} field.
+ * <p>
+ * When processing an item, single execution across all applications is guaranteed by setting the
+ * {@link DeadLetterEntry#getProcessingStarted()} property, locking other processes out of the sequence for the
+ * configured {@code claimDuration} (30 seconds by default).
+ * <p>
+ * The stored {@link DeadLetterEntry entries} are converted to a {@link JpaDeadLetter} when they need to be processed or
+ * filtered. In order to restore the original {@link EventMessage} a matching {@link DeadLetterJpaConverter} is used.
+ * The default supports all {@code EventMessage} implementations provided by the framework. If you have a custom
+ * variant, you have to build your own.
+ * <p>
+ * {@link org.axonframework.serialization.upcasting.Upcaster upcasters} are not supported by this implementation, so
+ * breaking changes for events messages stored in the queue should be avoided.
+ *
+ * @param <M> An implementation of {@link Message} contained in the {@link DeadLetter dead-letters} within this queue.
+ * @author Mitchell Herrijgers
+ * @since 4.6.0
+ */
+public class JpaSequencedDeadLetterQueue<M extends EventMessage<?>> implements SequencedDeadLetterQueue<M> {
+
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final String processingGroup;
+    private final EntityManagerProvider entityManagerProvider;
+    private final List<DeadLetterJpaConverter<EventMessage<?>>> converters;
+    private final TransactionManager transactionManager;
+    private final int maxSequences;
+    private final int maxSequenceSize;
+    private final int queryPageSize;
+    private final Serializer serializer;
+    private final Duration claimDuration;
+
+    protected <T extends EventMessage<?>> JpaSequencedDeadLetterQueue(Builder<T> builder) {
```

Comment:
We have JavaDoc on all `protected` constructors ingesting a `Builder`, so I'd expect some on this one too.

Reply:
Added :)